python/src/scipp/_bins.py

```diff
@@ -0,0 +1,196 @@
+        """
+        return _call_cpp_func(_cpp.buckets.sum, self._obj)
+
+    def size(self):
+        """Number of events or elements in a bin.
+
+        :return: The number of elements in each of the input bins.
+        """
+        return _call_cpp_func(_cpp.bin_size, self._obj)
+
+    def concatenate(self, other=None, dim=None, out=None):
+        """Concatenate bins element-wise by concatenating bin contents along
+        their internal bin dimension.
+
+        This can be used as a binary operation, or a reduction operation: The
+        bins to concatenate are either obtained element-wise from `self`
+        and `other`, or, if `dim` but not `other` is given, from all bins
+        along the given dimension.
+
+        :param other: Optional input containing bins.
+        :param dim: Optional dimension along which to merge bins. If not given
+                    and `other` is `None`, the bins are merged along all
+                    dimensions.
+        :param out: Optional output buffer.
+        :raises: If `other` is not binned data.
+        :return: The bins of the two inputs merged.
+        """
+        if other is not None and dim is not None:
+            raise RuntimeError(
+                "`concatenate` requires `other` or a `dim`, but not both.")
+        if other is not None:
+            if out is None:
+                return _call_cpp_func(_cpp.buckets.concatenate, self._obj,
+                                      other)
+            else:
+                if self._obj is out:
+                    _call_cpp_func(_cpp.buckets.append, self._obj, other)
+                else:
+                    out = _call_cpp_func(_cpp.buckets.concatenate, self._obj,
+                                         other)
+                return out
+        if out is not None:
+            raise RuntimeError("`out` arg not support for  along dim")
+        if dim is not None:
+            return _call_cpp_func(_cpp.buckets.concatenate, self._obj, dim)
+        raise RuntimeError("Reduction along all dims not supported yet.")
+
+
+class GroupbyBins:
+    """
+    Proxy for operations on bins of a groupby object
+    """
+    def __init__(self, obj):
+        self._obj = obj
+
+    def concatenate(self, dim):
+        return self._obj.concatenate(dim)
+
+
+def _bins(obj):
+    if _cpp.is_bins(obj):
+        return Bins(obj)
+    else:
+        return None
+
+
+def _groupby_bins(obj):
+    return GroupbyBins(obj)
+
+
+def bin(x, edges):
+    """Create binned data by binning data along all dimensions given by edges.
+
+    This does not histogram the data, each output bin will contain a "list" of
+    input values.
+
+    :return: Variable containing data in bins.
+    :seealso: :py:func:`scipp.histogram` for histogramming data,
+              :py:func:`scipp.bins` for creating binned data based on
+              explicitly given index ranges.
+    """
+    return _call_cpp_func(_cpp.bucketby, x, edges)
+
+
+def bins(*args, **kwargs):
```

Comment:
Is it dangerous for the users that `bin` and `bins` do something different?

Reply:
I don't think it is dangerous: The two functions have different keyword arguments, and we can even point out the hypothetical user typo in the exception message ("did you mean `sc.bin`").